spec/i18n/placeholders_spec.rb

```diff
@@ -0,0 +1,45 @@
+describe "Placeholders in strings" do
+  it "translations preserve placeholders in strings" do
+    dirs = [Rails.root.join('locale')]
+
+    Rails::Engine.subclasses.each do |engine|
+      dir = engine.root.join('locale')
+      dirs << dir if dir.exist?
+    end
+
+    errors = {}
+
+    dirs.each do |dir|
+      Pathname.glob(::File.join(dir, "**", "*.po")).each do |po_file|
+        po = FastGettext::PoFile.new(po_file)
+        locale = po_file.dirname.basename.to_s
+        next if locale == 'en'
+        po.data.keys.each do |key|
+          next if po.data[key].nil?
+          placeholders = key.scan /%{\w+}/
+          placeholders.sort!
+          next if placeholders.empty?
+          translated_placeholders = po.data[key].scan /%{\w+}/
+          translated_placeholders.sort!
+          if placeholders.uniq != translated_placeholders.uniq
+            errors[po_file.to_s] ||= []
+            errors[po_file.to_s] << [key, po.data[key]]
+          end
+        end
+      end
+    end
+
+    if errors.present?
+      errors.keys.each do |file|
+        puts ">> #{file}\n"
+        errors[file].each do |items|
+          puts items[0]
+          puts items[1]
+          puts
+        end
+        puts
+      end
```

Comment:
2 points here.

- Prefer blowing apart `items` with descriptive names instead of `items[0], items[1]`.  The latter is not readable.  So something like

    ```ruby
    errors[file].each do |key, value|
    ```

    Note that this is not even necessary if you use nested hashes because it's more natural to name both parts anyway.

- I'm not sure why you are decomposing iterating over a hash...it's not necessary to loop over the key and then looking up the values independently...This whole thing seems cleaner written as

```ruby
    # if errors.present?  NOTE: You also don't even need this line because iterating over an empty array is a no-op
      errors.each do |file, file_errors|
        puts ">> #{file}\n"
        file_errors.each do |key, value|  # I only put key value here because I don't know what they are... Use descriptive names.
          puts key
          puts value
          puts             # Incidentally, this can all be written as a single line... puts key, value, ""
        end
        puts
      end
    end
```